### PR TITLE
Makes testable and add the ability to pre-fill configuration values

### PR DESCRIPTION
**Pull request by aurium**

### CODE REVIEW

**Comment by aurium** [comment]
Hi @alarner! Is this fine?

**Comment by alarner** [comment]
@aurium sorry it's taken me so long to get to this. I"m taking a look at it tonight but it's taking longer than I expected because my old code was so bad :) I'm trying to remember how this thing works and also trying to understand the changes that you made. Thank you sooooo much for adding tests. That's making my life a whole lot easier right now. Hopefully I'll be able to merge this week.

File: index.js [hunk 1]
```diff
@@ -9,20 +9,60 @@ let validator = require('validator');
...
+					returnObject[key] = appendExtraData(tmpl[key], extra[key]);
+			} else {
+				if (!tmpl.hasOwnProperty(key)) returnObject[key] = extra[key];
+			}
+		}
+	}
+	return returnObject;
+}
+
+function setDefaultValues(parsedTmpl, values) {
```
Comment (alarner) [inline index.js:38]: Can we re-write this function like this?

``` js
function setDefaultValues(parsedTmpl, values) {
    for (let i=0; i<parsedTmpl.lines.length; i++) {
        const line = parsedTmpl.lines[i];
        let obj = values;
        for(let j=0; j<line.path.length; j++) {
            obj = obj[line.path[j]];
        }
        line.value = obj.toString();
    }
}
```

It's a little bit more code but I think it's more clear what's going on than with multiple variables changing inside each for loop.

File: index.js [hunk 2]
```diff
@@ -9,20 +9,60 @@ let validator = require('validator');
...
 		counter: counter
 	};
+
 	if(!_.isObject(obj)) {
 		let match = null;
 		let type = 'string';
 		
-		if(!_.isString(obj)) {
-			obj = obj.toString();
+		if(!_.isString(obj) || !obj.match(helpRegex)) {
```
Comment (alarner) [inline index.js:64]: It seems like this work is more suited to be done in the appendExtraData function. This way our template doesn't have a mix of instructions strings (eg. '[string] Defaults to "The environment."') and value strings (eg. 'The environment.')

File: test/config-template.test.js [hunk 3]
```diff
@@ -0,0 +1,353 @@
+var should = require('chai').should();
```
Comment (alarner) [inline test/config-template.test.js:1]: These tests are so wonderful. Thank you!

**Review by alarner — CHANGES_REQUESTED** [review]
Overall this is fantastic work. Just a few requests based on my understanding of what you've done here. If I'm misunderstanding something let me know and we can discuss. If you're too busy / don't want to make these changes just let me know and I can finish things up. I'm so excited to merge this in!

File: index.js (hunk 1 above)
Comment (aurium) [reply index.js:38]: I believe it is ok. Readability is important too!

File: index.js (hunk 2 above)
Comment (aurium) [reply index.js:64]: This place is better to allow poorly made templates to work fine.
Also, any JSON file can be used as template, and will respect the data type. So you are open to new use-cases like use this lib as a system tool to edit any JSON file. :grinning:

File: test/config-template.test.js (hunk 3 above)
Comment (aurium) [reply test/config-template.test.js:1]: oh stop it, you

**Pull request merged**